---
title: "TK_54: Analysis of scRNAseq data from human/mouse adipose tissue"
output: html_notebook
author: "Hernan Lorenzi"
date: "2023-05-08"
link-citations: TRUE
---

Goal: to utilize a published RNA seq dataset for extracting information about the expression of some genes which are relevant to my project. It will be great if you able to help me to extract data. If you agree, I can send the article and talk more about my project.

I'm mainly interested in the purinergic receptors expression (P2Y, P2X and adenosine receptors) in both human and mouse if it's possible to extract.

I'm interested in both bulkRNAseq and scRNAseq data. For scRNAseq I'm specifically interested in both human and mouse adipocytes subclusters, adipose stem and progenitor cell subclusters, and immune subclusters. I need a comparison between healthy and obese human individuals as well as lean and obese male mice (no need any female mice data).

### Load required libraries

```{r}
library(Seurat)
library(tidyverse)
library(cowplot)
```

### Load data from manuscript

[Manuscript reference](https://www.nature.com/articles/s41586-022-04518-2) [data source](https://singlecell.broadinstitute.org/single_cell/study/SCP1376/a-single-cell-atlas-of-human-and-mouse-white-adipose-tissue)

```{r}
tissue_names <- c("human_adipocytes", "human_ASPCs", "human_immune", "mouse_adipocytes", "mouse_ASPCs", "mouse_immune")
#tissue_names <- c("human_immune")

for (my_tissue in tissue_names){
  
  print(paste("Processing",my_tissue))
  
  my_tissue_obj <- read_rds(paste0("./data/",my_tissue,".rds"))

  ### Plot UMAP clusters
  if (my_tissue %in% c("human_immune","mouse_immune")){
    p.umap <- DimPlot(my_tissue_obj, reduction = "umap", ) +
    guides(colour=guide_legend(ncol=2, override.aes = list(size = 5)))
  } else {
    p.umap <- DimPlot(my_tissue_obj, reduction = "umap" )
  }
  ggsave2(filename = paste0("clusters_",my_tissue,"_umap.png"), 
          plot = p.umap, 
          width = 6, height = 4)
  
  features = c(rownames(my_tissue_obj@assays$RNA@data)[grep("P2RX",
                                                            rownames(my_tissue_obj@assays$RNA@data), 
                                                            ignore.case = T  )], 
               rownames(my_tissue_obj@assays$RNA@data)[grep("P2RY",
                                                            rownames(my_tissue_obj@assays$RNA@data), 
                                                            ignore.case = T  )],
               rownames(my_tissue_obj@assays$RNA@data)[grep("ADORA1|ADORA2B|ADORA3",
                                                            rownames(my_tissue_obj@assays$RNA@data), 
                                                            ignore.case = T  )]
              )
              
  p2 <- FeaturePlot(my_tissue_obj, features = features, pt.size=0.5, order = T, reduction = "umap")
  ggsave2(filename = paste0("gene_expression_",my_tissue,"_umap.png"), 
        plot = p2 + p.umap, 
        width = 20, height = 20)
  #print(p2 + p.umap)
}  
```

```{r}
sessionInfo()
```